### ISSUE TITLE
Support Lambda Layers

### ISSUE THREAD

**Issue by popey456963** [terraform-providers/terraform-provider-aws#6651]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS has recently released a new feature, "[Lambda Layers](https://aws.amazon.com/blogs/aws/new-for-aws-lambda-use-any-programming-language-and-share-common-components/)", a way of sharing code between Lambda functions.  It would be useful for Terraform to support this.

### New or Affected Resource(s)

* aws_lambda

### Potential Terraform Configuration

```hcl
# An implementation would likely have two aspects to it, an addition lambda parameter:

resource "aws_lambda_function" "example_function" {
  layers = [
    // reference to layer
    // another reference
  ]
}

# And the layer type itself

resource "aws_lambda_layer" "example_layer" {
  # layer metadata
  layer_name = ""
  tags = []
  description = ""

  # location of layer
  s3_bucket = []
  s3_key = []
  s3_object_version = []
}
```

**Comment by robertpeteuil** [comment]
I was going to submit the same request.  I'm glad to see someone has already suggested it.

**Comment by robertpeteuil** [comment]
Might also need a `data` resource to be able to retrieve the latest version of a layer.

It looks like the [GetLayerVersion](https://docs.aws.amazon.com/lambda/latest/dg/API_GetLayerVersion.html) API supports this.

**Comment by acburdine** [comment]
I'll take a look at this one 😄

**Comment by ispulkit** [comment]
Waiting for this to come to life...

**Comment by popey456963** [comment]
@acburdine Wondered if you'd had a chance to look into this yet?

**Comment by ampopp04** [comment]
If I wanted to deploy a simple AWSLambda-Python37-SciPy1x layer to a python lambda, is there a way to do it within terraform until your enhancement gets pushed? I can look through the docs to see if there is some sort of generic mechanism I can use to call a layer create. Great work by the way!

**Comment by zilman** [comment]
@acburdine Thank you for doing this, your PR (#6782) looks very straightforward, any idea if it will make it into 0.12.0?

**Comment by alixaxel** [comment]
When can we expect this to be included in the stable Terraform release?

**Comment by bflad** [comment]
Just to clear up any confusion, [Terraform core](https://github.com/hashicorp/terraform/blob/master/CHANGELOG.md) and the [Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md) are on independently versioned and released. The provider aims for a weekly release cadence, but there will be a gap in that schedule with the end of December holiday season, resuming first week of January.

As for when "this" will be released, depends on what you're looking for:
* A new `aws_lambda_layer_version` resource for managing a version of a Lambda layer is in progress with #6782. If review items are addressed (may require additional rounds) and documentation is added before mid Thursday, that can get released this week. If not, very early January.
* Once the above is merged, that opens the doors to creating other surrounding Lambda Layer resources (e.g. Lambda Layer Version Permissions), data source(s), and updating the `aws_lambda_function` resource itself. Unless those are made extremely quickly and without review items, they probably won't land until early January given the current timing.

**Comment by jonathan-kosgei** [comment]
#6782 Seems complete? @bflad

**Comment by bflad** [comment]
The review comments in #6782 are not fully addressed.

**Comment by nvdbleek** [comment]
@bflad sorry to bother you again with this issue. But is there an update about when we could use layers in terraform? It looks like @acburdine addressed the remarks on the PR. If possible I want to get a rough estimate so I know I can wait on it, or try to work around this using a Null provider.

**Comment by bflad** [comment]
The `aws_lambda_layer_version` resource has been merged in and will release with version 1.56.0 of the Terraform AWS provider in the next day or two.

**Comment by bflad** [comment]
Support for a `layers` argument in the `aws_lambda_function` resource and a `layers` attribute in the `aws_lambda_function` data source has also been merged for release with version 1.56.0 of the Terraform AWS provider.

While basic management use cases should now be covered, two pieces of functionality do still remain:
* A new `aws_lambda_layer_version` data source to lookup Lambda Layers
* A new `aws_lambda_layer_version_permission` resource (similar to `aws_lambda_permission`) for allowing cross-account access

**Comment by bflad** [comment]
@acburdine I think we can omit the `most_recent` argument and just default to that behavior if `version` is not provided. 😄

**Comment by acburdine** [comment]
sounds good - will work on that as well 👍

**Comment by bflad** [comment]
The new `aws_lambda_layer_version` resource and `layers` support in `aws_lambda_function` resource/data source has been released in [version 1.56.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1560-january-16-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by oxlade39** [comment]
@bflad it doesn't seem to support data. 

I need to lookup a layer in my lambda deployment. How would I do that?

**Comment by acburdine** [comment]
@oxlade39 the data source hasn't been implemented yet. I've got some initial work on it locally that I've been meaning to get back to, will see if I can get it finished sometime this week or next.

**Comment by jhulfikarali** [comment]
@acburdine sorry to bother again with the same question. Is the work for data support is done? If so can you point me to the documentation to it?

**Comment by kkpedrok** [comment]
How do I add "AWSLambda-Python36-SciPy1x" to my deploy? That is a layer that already exists in AWS...

**Comment by oxlade39** [comment]
@kkpedrok Until the data support is added you will have to manually look up the arn for your region from the aws console and then use directly in your lambda terrraform. You probably want to have a property for the arn to make it simpler to change/upgrade

**Comment by acburdine** [comment]
Apologies for the delay, finally got the lambda_layer datasource submitted in #8577 😄

**Comment by bflad** [comment]
Hi everyone! 👋 Great news, a new `aws_lambda_layer_version` data source has been merged and will release later today with version 2.10.0 of the Terraform AWS Provider. 🎉  Thanks to @acburdine and others who made this functionality possible! 

Since it appears we have covered functionality requested in the initial feature request here and these general "support X" type feature requests do not have a great definition of "done", I'm going to opt to close this issue now so we can focus on specific asks going forward.

For any additional feature requests or bug reports with the Lambda Layer functionality in the Terraform AWS Provider, please create [new GitHub issues](https://github.com/terraform-providers/terraform-provider-aws/issues/new/choose) so we can triage them. Thanks!

**Comment by acburdine** [comment]
@bflad the lambda_layer_permission resource hasn't been implemented yet, which was I believe in this issue as well. Can open a new issue though if needed

**Comment by bflad** [comment]
Let's treat it separately now 👍

**Comment by bflad** [comment]
The `aws_lambda_layer_version` data source has been released in [version 2.10.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2100-may-10-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by svlaskovic** [comment]
> If I wanted to deploy a simple AWSLambda-Python37-SciPy1x layer to a python lambda, is there a way to do it within terraform until your enhancement gets pushed? I can look through the docs to see if there is some sort of generic mechanism I can use to call a layer create. Great work by the way!

Hi,

I am trying to achieve the same?

May you please let me know if you done the same and how?



Thank you,
S

**Comment by stasovon** [comment]
@svlaskovic @kpedrok the following worked for me:
```
resource "aws_lambda_function" "my_lambda" {
  layers = ["arn:aws:lambda:eu-west-1:399891621064:layer:AWSLambda-Python37-SciPy1x:2"]
  ...
}
```
I didn't create a aws_lambda_layer_version resource, the above was enough.
